### PR TITLE
feat: add IbetEscrow

### LINKED ISSUE #320: [FEATURE] Escrow contract
### 1. Is your feature request related to a problem? Please describe.

- I'd like to add a simple contract for escrowing tokens.

### 2. Describe the solution you'd like

- The specifications are as follows

### State

- balances（コントラクト入庫残高）
```solidity
// account => token => amount
mapping(address => mapping(address => uint256)) private balances;
```

- commitments（エスクロー中数量：拘束数量）
```solidity
// account => token => amount
mapping(address => mapping(address => uint256)) private commitments;
```

- latest escrow id（直近エスクローID）
```solidity
uint256 public latestEscrowId
```

### Methods

- token fallback
```solidity
/// @param _from アカウントアドレス：残高を保有するアドレス
/// @param _value デポジット数量
function tokenFallback(address _from, uint _value, bytes memory /*_data*/) public
```

- withdraw
```solidity
/// @param _token トークンアドレス
/// @return 処理結果
function withdraw(address _token)
```

- create escrow
```solidity
/// @param _token トークンアドレス
/// @param _recipient トークン受領者
/// @param _amount 数量
/// @param _agent エスクローエージェント
/// @param _data イベント出力用の任意のデータ
function createEscrow(
    address _token,
    address _recipient,
    uint256 _amount,
    address _agent,
    string memory _data
)
```

- cancel escrow
```solidity    
/// @param _escrowId エスクローID
function cancelEscrow(uint256 _escrowId)
```

- finish escrow
```solidity
/// @param _escrowId エスクローID
function finishEscrow(uint256 _escrowId)
```

### Events

- escrow created
```solidity
event EscrowCreated(
    uint256 indexed escrowId,
    address indexed token,
    address sender,
    address recipient,
    uint256 amount,
    address agent,
    string data
);
```

- escrow canceled
```solidity
event EscrowCanceled(
    uint256 indexed escrowId,
    address indexed token,
    address sender,
    address recipient,
    uint256 amount,
    address agent
);
```

- escrow finished
```solidity
event EscrowFinished(
    uint256 indexed escrowId,
    address indexed token,
    address sender,
    address recipient,
    uint256 amount,
    address agent
);
```

- withdrawn
```solidity
event Withdrawn(
    address indexed token,
    address indexed account
);
```

### PR DESCRIPTION
**Pull request by YoshihitoAso**
close #320

### CODE REVIEW

File: contracts/exchange/ibetEscrow.sol [hunk 1]
```diff
@@ -0,0 +1,422 @@
+/**
```
Comment (kyken) [inline contracts/exchange/ibetEscrow.sol:1]: [IMO]If you follow convention, fileName might be UpperCamel.

File: contracts/exchange/ibetEscrow.sol [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        );
+
+        Escrow memory escrow;
+        (escrow.token, escrow.sender, escrow.recipient,
+            escrow.amount, escrow.agent, escrow.status) =
+                EscrowStorage(storageAddress).getEscrow(_escrowId);
+
+        // チェック：エスクローが有効であること
+        require(
+            escrow.status == false,
```
Comment (kyken) [inline contracts/exchange/ibetEscrow.sol:253]: [IMO]`status` may be difficult to understand...

**Review by kyken — COMMENTED** [review]
left some comments

File: contracts/exchange/ibetEscrow.sol [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+    /// @param _value デポジット数量
+    function tokenFallback(address _from, uint _value, bytes memory /*_data*/)
+        public
+        override
+    {
+        setBalance(
+            _from,
+            msg.sender,
+            balanceOf(_from, msg.sender).add(_value)
+        );
```
Comment (kyken) [inline contracts/exchange/ibetEscrow.sol:404]: [IMO]i suspect that Event emit is needed after tokenFallback.

File: contracts/exchange/ibetEscrow.sol [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+    constructor(address _storageAddress)
+    {
+        storageAddress = _storageAddress;
+    }
+
+    // ---------------------------------------------------------------
+    // Function: Storage
+    // ---------------------------------------------------------------
+
+    struct Escrow {
```
[the remark this answers is not in the document]
Comment (YoshihitoAso) [reply contracts/exchange/ibetEscrow.sol:90]: Unfortunately, it is not possible to access other contract struct.

File: contracts/exchange/ibetEscrow.sol [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+            _from,
+            msg.sender,
+            balanceOf(_from, msg.sender).add(_value)
+        );
+    }
+
+    /// @notice アドレスがコントラクトアドレスであるかを判定
+    /// @param _addr アドレス
+    /// @return is_contract 判定結果
+    function isContract(address _addr)
```
[the remark this answers is not in the document]
Comment (YoshihitoAso) [reply contracts/exchange/ibetEscrow.sol]: That's right. I'll delete it.

File: contracts/exchange/ibetEscrow.sol (hunk 1 above)
Comment (YoshihitoAso) [reply contracts/exchange/ibetEscrow.sol]: oops

File: contracts/exchange/ibetEscrow.sol (hunk 2 above)
Comment (YoshihitoAso) [reply contracts/exchange/ibetEscrow.sol]: You're right, and I did miss the logic.
I will fix it.

File: contracts/exchange/ibetEscrow.sol (hunk 3 above)
Comment (YoshihitoAso) [reply contracts/exchange/ibetEscrow.sol]: You're right. I will add a new event.

**Pull request merged**